include/hip/hcc_detail/hip_atomic.h

```diff
@@ -74,18 +74,10 @@ float atomicAdd_impl(float* address, float val)
     return __uint_as_float(r);
 }
 __device__
+__attribute__((const))
 inline
-bool __hip_is_shared(const __attribute__((address_space(0))) void* ptr) noexcept
-{   // TODO: this is ersatz for __builtin_amdgcn_is_shared.
-    #if defined(__HIP_DEVICE_COMPILE__)
-        const unsigned int gp = reinterpret_cast<unsigned long long>(ptr);
-
-        return
-            gp ==(__builtin_amdgcn_s_getreg((15 << 11) | (16 << 6) | 15) << 16);
-    #else
-        return false;
-    #endif
-}
+bool __hip_is_shared(const __attribute__((address_space(0))) void*) noexcept
+    asm("llvm.amdgcn.is.shared");
```

Comment:
why we cannot use builtin in HCC? If flat-addrspace is missing in HCC, it should enable it. Directly using LLVM IR intrinsics should not be encouraged considering they are not the interface at language level.